File: homework-1/src/validators/transactionValidator.ts
import { CreateTransactionInput, TransactionType, ValidationError } from '../models/transaction';

/**
 * ISO 4217 currency codes supported by the system
 */
const VALID_CURRENCIES = [
  'USD', 'EUR', 'GBP', 'JPY', 'CHF', 'CAD', 'AUD', 'NZD',
  'CNY', 'INR', 'BRL', 'ZAR', 'SEK', 'NOK', 'DKK', 'PLN',
  'RUB', 'MXN', 'SGD', 'HKD', 'KRW', 'TRY', 'AED', 'SAR'
];

/**
 * Account number format: ACC-XXXXX where X is alphanumeric
 */
const ACCOUNT_FORMAT_REGEX = /^ACC-[A-Z0-9]{5}$/;

/**
 * Validates a transaction input and returns validation errors if any
 */
export function validateTransaction(input: CreateTransactionInput): ValidationError[] {
  const errors: ValidationError[] = [];

  // Validate amount
  if (input.amount === undefined || input.amount === null) {
    errors.push({
      field: 'amount',
      message: 'Amount is required'
    });
  } else if (typeof input.amount !== 'number' || isNaN(input.amount)) {
    errors.push({
      field: 'amount',
      message: 'Amount must be a valid number'
    });
  } else if (input.amount <= 0) {
    errors.push({
      field: 'amount',
      message: 'Amount must be a positive number'
    });
  } else if (!isValidDecimalPlaces(input.amount, 2)) {
    errors.push({
      field: 'amount',
      message: 'Amount must have maximum 2 decimal places'
    });
  }

  // Validate fromAccount
  if (!input.fromAccount) {
    errors.push({
      field: 'fromAccount',
      message: 'From account is required'
    });
  } else if (!isValidAccountFormat(input.fromAccount)) {
    errors.push({
      field: 'fromAccount',
      message: 'From account must follow format ACC-XXXXX (where X is alphanumeric)'
    });
  }

  // Validate toAccount
  if (!input.toAccount) {
    errors.push({
      field: 'toAccount',
      message: 'To account is required'
    });
  } else if (!isValidAccountFormat(input.toAccount)) {
    errors.push({
      field: 'toAccount',
      message: 'To account must follow format ACC-XXXXX (where X is alphanumeric)'
    });
  }

  // Validate currency
  if (!input.currency) {
    errors.push({
      field: 'currency',
      message: 'Currency is required'
    });
  } else if (!isValidCurrency(input.currency)) {
    errors.push({
      field: 'currency',
      message: `Invalid currency code. Must be one of: ${VALID_CURRENCIES.join(', ')}`
    });
  }

  // Validate type
  if (!input.type) {
    errors.push({
      field: 'type',
      message: 'Transaction type is required'
    });
  } else if (!isValidTransactionType(input.type)) {
    errors.push({
      field: 'type',
      message: `Transaction type must be one of: ${Object.values(TransactionType).join(', ')}`
    });
  }

  // Business logic validation
  if (input.type === TransactionType.DEPOSIT && input.fromAccount !== 'ACC-00000') {
    // For deposits, fromAccount should typically be a system account
    // This is a soft validation - we'll allow it but could be stricter
  }

  if (input.type === TransactionType.WITHDRAWAL && input.toAccount !== 'ACC-00000') {
    // For withdrawals, toAccount should typically be a system account
    // This is a soft validation - we'll allow it but could be stricter
  }

  return errors;
}

/**
 * Validates that a number has at most the specified decimal places
 */
function isValidDecimalPlaces(value: number, maxDecimals: number): boolean {
  const decimalPart = value.toString().split('.')[1];
  return !decimalPart || decimalPart.length <= maxDecimals;
}

/**
 * Validates account format: ACC-XXXXX
 */
function isValidAccountFormat(account: string): boolean {
  return ACCOUNT_FORMAT_REGEX.test(account);
}

/**
 * Validates ISO 4217 currency code
 */
function isValidCurrency(currency: string): boolean {
  return VALID_CURRENCIES.includes(currency.toUpperCase());
}

/**
 * Validates transaction type
 */
function isValidTransactionType(type: string): boolean {
  return Object.values(TransactionType).includes(type as TransactionType);
}

/**
 * Validates date format (ISO 8601)
 */
export function isValidDateFormat(dateString: string): boolean {
  const date = new Date(dateString);
  return !isNaN(date.getTime());
}

File: homework-6/agents/transaction_validator.py
"""Transaction Validator Agent.

Validates transaction message envelopes for:
- Required fields presence
- Positive Decimal amount
- ISO 4217 currency whitelist membership
"""
import argparse
import json
import logging
import uuid
from datetime import datetime, timezone
from decimal import Decimal, InvalidOperation
from pathlib import Path

AGENT_NAME = "transaction_validator"

VALID_CURRENCIES: frozenset = frozenset({
    "USD", "EUR", "GBP", "JPY", "CAD", "CHF", "AUD", "CNY",
    "SEK", "NOK", "DKK", "NZD", "SGD", "HKD", "MXN", "BRL",
    "INR", "KRW", "ZAR", "TRY",
})

REQUIRED_FIELDS = (
    "transaction_id",
    "amount",
    "currency",
    "source_account",
    "destination_account",
    "timestamp",
)

logger = logging.getLogger(AGENT_NAME)


def _mask_account(account_id: str) -> str:
    """Return a PII-safe masked version of an account identifier."""
    if not account_id:
        return "***"
    parts = account_id.split("-", 1)
    if len(parts) == 2:
        return f"{parts[0]}-****"
    return "****"


class TransactionValidator:
    """Validates transaction message envelopes."""

    def process_message(self, message: dict) -> dict:
        """Validate a transaction and return an updated message envelope.

        Returns envelope with status='validated' targeting fraud_detector,
        or status='rejected' with rejection_reason targeting results.
        """
        data = message.get("data", {})
        txn_id = data.get("transaction_id", "UNKNOWN")
        src = _mask_account(data.get("source_account", ""))
        dst = _mask_account(data.get("destination_account", ""))
        log_prefix = f"txn={txn_id} src={src} dst={dst}"

        # 1. Required fields
        for field in REQUIRED_FIELDS:
            value = data.get(field)
            if value is None or value == "":
                logger.info(
                    "%s | REJECTED | reason=MISSING_FIELD field=%s",
                    log_prefix, field,
                )
                return self._build_result(
                    message, "rejected", "results",
                    rejection_reason="MISSING_FIELD",
                    rejection_detail=f"Missing required field: {field}",
                )

        # 2. Amount must be a positive Decimal
        try:
            amount = Decimal(str(data["amount"]))
        except InvalidOperation:
            logger.info("%s | REJECTED | reason=INVALID_AMOUNT", log_prefix)
            return self._build_result(
                message, "rejected", "results",
                rejection_reason="INVALID_AMOUNT",
                rejection_detail="Amount is not a valid decimal number",
            )

        if amount <= 0:
            logger.info(
                "%s | REJECTED | reason=INVALID_AMOUNT amount=%s",
                log_prefix, amount,
            )
            return self._build_result(
                message, "rejected", "results",
                rejection_reason="INVALID_AMOUNT",
                rejection_detail=f"Amount must be positive, got {amount}",
            )

        # 3. Currency must be in ISO 4217 whitelist
        currency = str(data.get("currency", "")).strip().upper()
        if currency not in VALID_CURRENCIES:
            logger.info(
                "%s | REJECTED | reason=INVALID_CURRENCY currency=%s",
                log_prefix, currency,
            )
            return self._build_result(
                message, "rejected", "results",
                rejection_reason="INVALID_CURRENCY",
                rejection_detail=f"Currency '{currency}' is not in the ISO 4217 whitelist",
            )

        logger.info(
            "%s | VALIDATED | amount=%s currency=%s", log_prefix, amount, currency,
        )
        return self._build_result(
            message, "validated", "fraud_detector",
            amount=str(amount),
        )

    def _build_result(
        self,
        message: dict,
        status: str,
        target_agent: str,
        **data_overrides,
    ) -> dict:
        result = dict(message)
        result["source_agent"] = AGENT_NAME
        result["target_agent"] = target_agent
        result["timestamp"] = datetime.now(timezone.utc).isoformat()
        result["data"] = dict(message.get("data", {}))
        result["data"]["status"] = status
        result["data"].update(data_overrides)
        return result


def _dry_run(input_path: Path) -> None:
    """Validate transactions in dry-run mode and print a summary table."""
    with open(input_path) as f:
        transactions = json.load(f)

    validator = TransactionValidator()
    rows = []
    for txn in transactions:
        message = {
            "message_id": str(uuid.uuid4()),
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "source_agent": "integrator",
            "target_agent": AGENT_NAME,
            "message_type": "transaction",
            "data": txn,
        }
        result = validator.process_message(message)
        d = result["data"]
        rows.append((
            d.get("transaction_id", "?"),
            d["status"],
            d.get("rejection_reason", ""),
        ))

    valid_count = sum(1 for _, s, _ in rows if s == "validated")
    invalid_count = len(rows) - valid_count

    print(f"\n{'='*60}")
    print(f"{'Transaction Validation Dry Run':^60}")
    print(f"{'='*60}")
    print(f"{'TXN ID':<12} {'STATUS':<15} REASON")
    print("-" * 60)
    for txn_id, status, reason in rows:
        print(f"{txn_id:<12} {status:<15} {reason}")
    print("=" * 60)
    print(f"Total: {len(rows)} | Valid: {valid_count} | Invalid: {invalid_count}")


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(name)s] %(message)s")
    parser = argparse.ArgumentParser(description="Transaction Validator Agent")
    parser.add_argument("--dry-run", action="store_true",
                        help="Validate transactions without writing files")
    parser.add_argument("--input", default="sample-transactions.json",
                        help="Input transaction JSON file")
    args = parser.parse_args()

    if args.dry_run:
        _dry_run(Path(args.input))
